pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,263 @@
+
+	// waitEnqueue shows how long a given event was waiting on the queue before
+	// it was actually processed.
+	waitEnqueue spanstat.SpanStat
+
+	// durationStat shows how long the actual processing of the event took. This
+	// is the time for how long Handle() takes for the event.
+	durationStat spanstat.SpanStat
+
+	// waitConsumeOffQueue shows how long it took for the event to be consumed
+	// off of the queue.
+	waitConsumeOffQueue spanstat.SpanStat
+}
+
+// NewEvent returns an Event with all fields initialized.
+func NewEvent(meta EventHandler) *Event {
+	return &Event{
+		Metadata:     meta,
+		eventResults: make(chan interface{}, 1),
+		cancelled:    make(chan struct{}),
+		stats:        eventStatistics{},
+	}
+}
+
+// WasCancelled returns whether the cancelled channel for the given Event has
+// been closed or not. Cancellation occurs if the event was not processed yet
+// by an EventQueue onto which this Event was Enqueued, and the queue is closed,
+// or if the event was attempted to be scheduled onto an EventQueue which has
+// already been closed.
+func (ev *Event) WasCancelled() bool {
+	select {
+	case <-ev.cancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+// Enqueue pushes the given event onto the EventQueue. If the queue has been
+// stopped, the Event will not be enqueued, and its cancel channel will be
+// closed, indicating that the Event was not ran. This function may block if
+// the queue is at its capacity for events.
+func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
+
+	if ev == nil {
+		return nil
+	}
+
+	// Track that event has been Enqueued.
+	q.closeWaitGroup.Add(1)
+	defer q.closeWaitGroup.Done()
+
+	select {
+	// The event should be drained from the queue (e.g., it should not be
+	// processed).
+	case <-q.drain:
+		// Closed eventResults channel signifies cancellation.
+		close(ev.cancelled)
+		close(ev.eventResults)
+
+		return ev.eventResults
+	default:
+		// The events channel may be closed even if an event has been pushed
+		// onto the events channel, as events are consumed off of the events
+		// channel asynchronously! If the EventQueue is closed before this
+		// event is processed, then it will be cancelled.
+
+		ev.stats.waitEnqueue.Start()
+		q.events <- ev
+		ev.stats.waitEnqueue.End(true)
+		ev.stats.waitConsumeOffQueue.Start()
+		return ev.eventResults
+	}
+}
+
+func (ev *Event) printStats() {
+	log.WithFields(logrus.Fields{
+		"eventType":                    reflect.TypeOf(ev.Metadata).String(),
+		"eventHandlingDuration":        ev.stats.durationStat.Total(),
+		"eventEnqueueWaitTime":         ev.stats.waitEnqueue.Total(),
+		"eventConsumeOffQueueWaitTime": ev.stats.waitConsumeOffQueue.Total(),
+	}).Debug("EventQueue event processing statistics")
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
```

Comment:
nit: I don't think this assumption still holds?

Reply:
I thought I had caught this one, sorry for the noise. Will update.